### PR TITLE
124 validate postcodes

### PR DESCRIPTION
**Pull request by jwestw**
## Code Changes in `.github/workflows/pytest-action.yaml`

In the `.github/workflows/pytest-action.yaml` file:

- I added the `working-directory` field to specify the working directory as the root of the repository.
- I modified the `coverage run` command to include the `--ignore=tests/test_utils/test_hdfs_mods.py` option. This option excludes the specified test file from the coverage analysis.

## Code Changes in `environment.yml`

In the `environment.yml` file:

- I changed the Python version from `python=3` to `python=3.6`.
- I added the `more-itertools` package with version `5.0.0` as a dependency.
- I added the `postcodes_uk` package as a dependency.

## Code Changes in `src/data_ingest/loading.py`

In the `src/data_ingest/loading.py` file:

- I added the import statement for the `Tuple` type from the `typing` module.
- I defined a new function called `parse_snap_data` which takes in a `snapdata` dictionary (with a default value of `snapshot_path`) and returns a tuple of two pandas DataFrames. This function loads the data from the survey via the SPP snapshot and parses it into separate dataframes for survey contributors (company details) and their responses.
- I modified the previous code that directly loaded the `snapdata` using `hdfs_load_json` to now call the `parse_snap_data` function.
- Inside the `parse_snap_data` function, I added type hints to the function signature and added docstrings to describe the function's purpose and arguments.
- I replaced the direct assignment of `contributerdict` and `responsesdict` with the corresponding lines inside the `parse_snap_data` function.
- I modified the previous code that created the `contributers` and `responses` dataframes to now use the variable names `contributers_df` and `responses_df`.
- I added a return statement at the end of the `parse_snap_data` function to return the two dataframes.

## Code Changes in `src/data_validation/validation.py`

In the `src/data_validation/validation.py` file:

- Added import statements for the necessary modules and functions.
- Updated the `get_masterlist` function to load the masterlist of postcodes from a CSV file.
- Added the `validate_post_col` function to check if all postcodes in a specified DataFrame column are valid UK postcodes.
- Added the `validate_postcode_pattern` function to validate UK postcodes based on their pattern.
- Implemented logging for warning messages related to unreal and invalid pattern postcodes.
- Combined the lists of unreal and invalid pattern postcodes and raised a `ValueError` if any invalid postcodes are found.
- Finally, I added a return statement in the `validate_post_col` function to return `True` if all postcodes are valid and real.

## Code Changes in `src/main.py`

In the `src/main.py` file:

- Added imports for the `loading` module from `src.data_ingest` and the `validation` module from `src.data_validation`.
- Added the `masterlist_path` variable to store the path from the config settings.
- Updated the `run_pipeline` function to include the data ingest and data validation steps.
- Added print statements to display sample data from the data ingest step.
- Added a call to `validate_post_col` function from the `validation` module to validate the postcode column in the `contributers_df` DataFrame.

## Code Changes in `tests/README.md`

In the `tests/README.md` file:

- I modified the description of the `tests` folder overview to include the requirement of storing each test module in its correct folder.

# Instructions for the tester

* install new requirements from `environment.yml`
* You will need the masterlist of postcodes which I used - ask me for that

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #124

#### Code

- [x] **Code runs** The code runs on my machine and/or CDSW
- [x] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [x] **Requirements** My/our code functions according to the requirements of the ticket
- [x] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used

- [x] **Clean Code**
    - [x] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [x] Code passes flake8 linting check
    - [x] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [x] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [x] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [x] Includes `Args` and `returns` for all major functions
    - [x] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [x] All data is excluded from this pull request
- [x] Secrets checker pre-commit passes

#### Testing
- [x] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: tests/test_validation.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+def test_validate_postcode():
+    # Valid postcodes
+    assert validate_postcode("AB12 3CD") == True
+    assert validate_postcode("DE34 5FG") == True
+    assert validate_postcode("HI67 8JK") == True
+
+    # Invalid postcodes
+    assert validate_postcode("EFG 456") == False
+    assert validate_postcode("HIJ 789") == False
+    assert validate_postcode("KL1M 2NO") == False
```
Comment (ali-i3) [inline tests/test_validation.py:57]: validate_postcode("KL1M 2NO") returns True so assertion fails

File: tests/test_validation.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    assert validate_postcode("B27 OAG") == False  # Zero is actually an "O"
+
+    # Edge cases
+    assert validate_postcode(None) == False  # None value should fail
+    assert validate_postcode("") == False  # Empty string
+    assert validate_postcode(" ") == False  # Whitespace
+    assert validate_postcode("AB123CD") == False  # Missing space - othewise valid
```
Comment (ali-i3) [inline tests/test_validation.py:64]: validate_postcode("AB123CD") returns True and assertion fails

File: tests/test_validation.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import pytest
+from src.data_validation import validate_postcode
+
+
+import pytest
+import pandas as pd
+from your_module import validate_post_col
```
Comment (ali-i3) [inline tests/test_validation.py:7]: Import issues here. Try: from `src.data_validation.validation` import `validate_postcode`, `validate_post_col`

File: tests/test_validation.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        validate_post_col(df_missing_column)
+    assert str(error.value) == "'referencepostcode'"  # Missing column
+
+    # Edge cases: missing DataFrame
+    df_missing_dataframe = None
+    with pytest.raises(AttributeError):
+        validate_post_col(df_missing_dataframe)  # Missing DataFrame
+
+    # Edge cases: empty reference postcode column
+    df_no_postcodes = pd.DataFrame({"referencepostcode": [""]})
```
Comment (ali-i3) [inline tests/test_validation.py:44]: This assertion fails since validate_post_col(df_no_postcodes) returns a ValueError. 
Try:
    df_no_postcodes = pd.DataFrame({"referencepostcode": [""]})
    with pytest.raises(ValueError):
        validate_post_col(df_no_postcodes)  # Empty postcode

File: tests/test_validation.py (hunk 2 above)
Comment (jwestw) [reply tests/test_validation.py]: Hmm this is an interesting test case.

File: tests/test_validation.py (hunk 4 above)
Comment (jwestw) [reply tests/test_validation.py]: I have changed this.

File: tests/test_validation.py (hunk 3 above)
Comment (jwestw) [reply tests/test_validation.py]: Yeah ooops. Thanks for pointing this one out. I had actually spotted it myself

File: tests/test_validation.py (hunk 2 above)
Comment (jwestw) [reply tests/test_validation.py]: Message from Ilyas (via Teams)

> I ran validation.is_valid_postcode('KL1M 2NO') using the uk-postcodes-utils package [uk-postcode-utils · PyPI](https://pypi.org/project/uk-postcode-utils/) and it returned false. Might be a more accurate validator.

I also tried this library, and while it worked on 'KL1M 2NO' it failed to validate 'SW1A 1AA' (which is the postcode of Buckingham Palace)

image

File: tests/test_validation.py (hunk 1 above)
Comment (jwestw) [reply tests/test_validation.py]: [Further investigation](https://en.wikipedia.org/wiki/Postcodes_in_the_United_Kingdom#Validation) shows that "KL1M 2NO" is an allowed pattern, but it is not a real postcode. So the pattern is true, but it won't exist in an address database. 

So I got hold "ONSPD" data which I believe contains all postcodes to test this. The longer postcodes like do not have spaces in them. I tried with the Buckingham Palace postcode again and it only validates without a space. 

image

In the case of "KL1M 2NO", this is a longer one too (but not a real postcode). Knowing that I needed to remove space in the longer postcodes, I tried with and without the space and it fails in both cases.  

image

To confirm this is a fake postcode I have put it into Google maps. 

image

File: src/data_validation/validation.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    return valid_bool
+
+
+def get_masterlist(masterlist_path) -> pd.Series:
+    """This function loads the masterlist of postcodes from a csv file
+
+    Returns:
+        pd.Series: The dataframe of postcodes
+    """
+    masterlist = pd.read_csv(masterlist_path, usecols=["pcd"]).squeeze()
```
Comment (ali-i3) [inline src/data_validation/validation.py:41]: Changed this since it was failing. Squeeze function converts the pcd column to pd.series.

File: src/data_validation/validation.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    Example:
+        >>> df = pd.DataFrame(
+            {"referencepostcode": ["AB12 3CD", "EFG 456", "HIJ 789", "KL1M 2NO"]})
+        >>> validate_post_col(df, "example-path/to/masterlist.csv"")
+        ValueError: Invalid postcodes found: ['EFG 456', 'HIJ 789']
+    """
+    if not isinstance(df, pd.DataFrame):
+        raise TypeError(f"The dataframe you are attempting to validate is {type(df)}")
+
+    if config["global"]["postcode_csv_check"]:
```
Comment (ali-i3) [inline src/data_validation/validation.py:69]: Added a switch like we discussed. It's very basic. Turning the switch to False means the unit tests will fail since they are designed to check the masterlist. The tests might need a little tweaking to incorporate the switch

File: src/developer_config.yaml [hunk 7]
```diff
@@ -11,6 +12,7 @@ runlog_writer:
 paths:
   logs_foldername: "testing_pydoop"
   snapshot_path: "/ons/rdbe_dev/snapshot-202012-002-fba5c4ba-fb8c-4a62-87bb-66c725eea5fd.json"
+  masterlist_path: "data/external/ONSPD_NOV_2022_UK.csv"
```
Comment (ali-i3) [inline src/developer_config.yaml:15]: Added path to masterlist since it was missing

File: src/data_validation/validation.py (hunk 6 above)
Comment (james-westwood) [reply src/data_validation/validation.py:69]: This is amazing. Thanks

File: src/developer_config.yaml (hunk 7 above)
Comment (jwestw) [reply src/developer_config.yaml:15]: Great

File: src/data_validation/validation.py (hunk 5 above)
Comment (jwestw) [reply src/data_validation/validation.py:41]: Nice

File: src/data_validation/validation.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    return valid_bool
+
+def get_masterlist(masterlist_path) -> pd.Series:
+    """This function loads the masterlist of postcodes from a csv file
+
+    Returns:
+        pd.Series: The dataframe of postcodes
+    """
+    
+    masterlist = (pd.Series
```
[the remark this answers is not in the document]
Comment (jwestw) [reply src/data_validation/validation.py]: Think this is resolved

**Review by ali-i3 — APPROVED** [review]
Nice!
